chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -716,3 +1161,102 @@ func (b *BitcoindNotifier) RegisterBlockEpochNtfn() (*chainntnfs.BlockEpochEvent
 		}, nil
 	}
 }
+
+// nodePeer is a peer connected to the RPC client's backend coupled with a
+// channel that will retrieve messages from it.
+type nodePeer struct {
+	*peer.Peer
+	msgChan chan interface{}
+}
+
+// connectToOutboundPeer connects to a node's peer and sets up the provided
+// listeners. It guarantees that the version handshake has completed.
+func connectToOutboundPeer(addr string, params *chaincfg.Params,
+	listeners peer.MessageListeners, msgChan chan interface{}) (*nodePeer, error) {
+
+	// We set up a listener for a `verack` message, which will signal us
+	// when the version handshake has completed.
+	verack := make(chan struct{})
+	listeners.OnVerAck = func(_ *peer.Peer, msg *wire.MsgVerAck) {
+		verack <- struct{}{}
+	}
+
+	peerCfg := &peer.Config{
+		ChainParams: params,
+		Listeners:   listeners,
+	}
+
+	p, err := peer.NewOutboundPeer(peerCfg, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Allow the peer to start exchanging messages on the established
+	// connection.
+	p.AssociateConnection(conn)
+
+	// Since the peer will reject all messages without establishing a
+	// version handshake first, we make sure it has succeeded.
+	select {
+	case <-verack:
+	case <-time.After(15 * time.Second):
+		// TODO: adjust timeout?
+		// The version handshake did not succeed within the timeout, so
+		// we can proceed to disconnect from the peer.
+		p.Disconnect()
+		return nil, errors.New("peer did not complete version " +
+			"handshake within 15 seconds")
+	}
+
+	return &nodePeer{p, msgChan}, nil
+}
+
+// getBlockFromPeer fetches the block with the given hash from the peer.
+func getBlockFromPeer(p *nodePeer, blockHash *chainhash.Hash) (*btcutil.Block, error) {
+	// Create a message requesting the peer to provide us with the block
+	// that has the hash above and send it.
+	getDataMsg := wire.NewMsgGetData()
+	inv := wire.NewInvVect(wire.InvTypeWitnessBlock, blockHash)
+	if err := getDataMsg.AddInvVect(inv); err != nil {
+		return nil, err
+	}
+
+	p.QueueMessageWithEncoding(getDataMsg, nil, wire.WitnessEncoding)
+
+	select {
+	case msg := <-p.msgChan:
+		switch msg.(type) {
+		case *btcutil.Block:
+			return msg.(*btcutil.Block), nil
+		case *wire.MsgNotFound:
+			// The peer was not able to find a block with the
+			// requested hash.
+			return nil, fmt.Errorf("block with hash %v not found",
+				blockHash)
+		}
+	case <-time.After(15 * time.Second):
```

Comment:
I think we should make this method a method on the `Notifier`, such that we can return in the case `<-notifier.quit`